drivers/input/inport/hardware.c

```diff
@@ -0,0 +1,418 @@
+#pragma alloc_text(PAGE, InPortInitializeMouse)
+#pragma alloc_text(PAGE, InPortStopMouse)
+#endif
+
+#define READ_MOUSE(DeviceExtension, Port) \
+    READ_PORT_UCHAR((DeviceExtension)->IoBase + (Port))
+
+#define WRITE_MOUSE(DeviceExtension, Port, Data) \
+    WRITE_PORT_UCHAR((DeviceExtension)->IoBase + (Port), (Data))
+
+/*
+ * NEC
+ */
+#define NEC_BM_DATA          0x00
+
+#define NEC_BM_CONTROL       0x04
+    #define NEC_INT_ENABLE       0x00
+    #define NEC_INT_DISABLE      0x10
+
+    #define NEC_READ_X_LOW       0x00
+    #define NEC_READ_X_HIGH      0x20
+    #define NEC_READ_Y_LOW       0x40
+    #define NEC_READ_Y_HIGH      0x60
+
+    #define NEC_INPUT_CAPTURE    0x00
+    #define NEC_INPUT_HOLD       0x80
+
+#define NEC_BM_CONFIG        0x06
+    #define NEC_PPI_INT_ENABLE   0x08
+    #define NEC_PPI_INT_DISABLE  0x09
+    #define NEC_PPI_HC_NO_CLEAR  0x0E
+    #define NEC_PPI_HC_CLEAR     0x0F
+    #define NEC_PPI_DEFAULT_MODE 0x93
+
+#define NEC_BM_INT_RATE      0x4002
+    #define NEC_RATE_120_HZ      0x00
+    #define NEC_RATE_60_HZ       0x01
+    #define NEC_RATE_30_HZ       0x02
+    #define NEC_RATE_15_HZ       0x03
+
+#define NEC_BM_HIRESO_BASE   (PUCHAR)0x61
+
+#define NEC_BUTTON_RIGHT    0x20
+#define NEC_BUTTON_LEFT     0x80
+
+/*
+ * Microsoft InPort
+ */
+#define MS_INPORT_CONTROL    0x00
+    #define INPORT_REG_BTNS    0x00
+    #define INPORT_REG_X       0x01
+    #define INPORT_REG_Y       0x02
+    #define INPORT_REG_MODE    0x07
+    #define INPORT_RESET       0x80
+
+#define MS_INPORT_DATA       0x01
+    #define INPORT_MODE_IRQ    0x01
+    #define INPORT_MODE_BASE   0x10
+    #define INPORT_MODE_HOLD   0x20
+
+#define MS_INPORT_SIGNATURE  0x02
+
+#define MS_BUTTON_MIDDLE   0x01
+#define MS_BUTTON_LEFT     0x02
+#define MS_BUTTON_RIGHT    0x04
+
+/*
+ * Logitech
+ */
+#define LOG_BM_DATA          0x00
+
+#define LOG_BM_SIGNATURE     0x01
+    #define LOG_SIGNATURE_BYTE 0xA5
+
+#define LOG_BM_CONTROL       0x02
+    #define LOG_ENABLE_IRQ     0x00
+    #define LOG_DISABLE_IRQ    0x10
+    #define LOG_READ_X_LOW     0x80
+    #define LOG_READ_X_HIGH    0xA0
+    #define LOG_READ_Y_LOW     0xC0
+    #define LOG_READ_Y_HIGH    0xE0
+
+#define LOG_BM_CONFIG        0x03
+    #define LOG_DEFAULT_MODE   0x90
+    #define LOG_CONFIG_BYTE    0x91
+
+#define LOG_BUTTON_RIGHT     0x20
+#define LOG_BUTTON_MIDDLE    0x40
+#define LOG_BUTTON_LEFT      0x80
+
+/* FUNCTIONS ******************************************************************/
+
+VOID
+NTAPI
+InPortDpcForIsr(
+    _In_ PKDPC Dpc,
+    _In_opt_ PDEVICE_OBJECT DeviceObject,
+    _Inout_ PIRP Irp,
+    _In_opt_ PVOID Context)
+{
+    PINPORT_DEVICE_EXTENSION DeviceExtension = DeviceObject->DeviceExtension;
+    ULONG InputDataConsumed;
+    INPORT_RAW_DATA RawData;
+    ULONG ButtonDiff;
+
+    UNREFERENCED_PARAMETER(Dpc);
+    UNREFERENCED_PARAMETER(Irp);
+    UNREFERENCED_PARAMETER(Context);
+
+    /* Copy raw data */
+    RawData = DeviceExtension->RawData;
```

Comment:
It's generally not safe to access the same data in the ISR and the DPC. Should this use `KeAcquireInterruptSpinLock` or similar?